### PR TITLE
Fix race conditions between CTRL+C and RunAsync in dotnet-watch

### LINKED ISSUE #288: [Flaky test] ProgramTests.ConsoleCancelKey
```
System.TimeoutException : Task exceeded max running time of 30s at C:\BuildAgent\work\2b0c8ce712895332\.r\DotNetTools\test\Microsoft.DotNet.Watcher.Tools.Tests\ProgramTests.cs:42
   at System.Threading.Tasks.TaskExtensions.<OrTimeout>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Threading.Tasks.TaskExtensions.<OrTimeout>d__0`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.DotNet.Watcher.Tools.Tests.ProgramTests.<ConsoleCancelKey>d__3.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

### PR DESCRIPTION
**Pull request by natemcmaster**
In slow machines, the CTRL+C cancellation token in this test may have fire before Program subscribed to the ConsoleCancelKey event. 

Also, in testing under load, this race condition could also throw ObjectDisposedException from _ctrlCTokenSource.Cancel(). This is caused because the lambda used to subscribe to ConsoleCancelKey captures _ctrlCTokenSource in a closure that is removed from the event.
 
The flakiness and ObjectDisposed bug is fixed by better controlling the lifetime of the cancellation token source and ensuring we app unsubscribes from ConsoleCancelKey before closing.

Also, cleaned out some duplicate code and reference Microsoft.AspNetCore.Testing instead.

Fixes #288 - cc @pranavkm

### CODE REVIEW

**Comment by natemcmaster** [comment]
Update: it turns out dotnet-run is the source our testing woes. Something in the latest SDK is slow to build + launch. Increasing timeouts seems to solve failing tests. This should fix #285. cc @mikeharder

File: src/Microsoft.DotNet.Watcher.Tools/CommandLineOptions.cs [hunk 1]
```diff
@@ -38,6 +38,10 @@ public static CommandLineOptions Parse(string[] args, IConsole console)
   changes. This is required for some file systems, such as network shares,
   Docker mounted volumes, and other virtual file systems.
 
+  DOTNET_WATCH
+  dotnet-watch sets this variable to '1' to make it possible to detect
+  when an process is a child process of dotnet-watch.
```
Comment (cesarbs) [inline src/Microsoft.DotNet.Watcher.Tools/CommandLineOptions.cs]: nit: a process

File: src/Microsoft.DotNet.Watcher.Tools/Program.cs [hunk 2]
```diff
@@ -60,54 +67,58 @@ public static int Main(string[] args)
...
+        {
+            if (!_cts.IsCancellationRequested)
+            {
+                _reporter.Output("Shutdown requested. Press Ctrl+C again to force exit.");
+                // suppress default result of CTRL+C
+                args.Cancel = true;
             }
+            else
+            {
+                args.Cancel = false;
```
Comment (cesarbs) [inline src/Microsoft.DotNet.Watcher.Tools/Program.cs]: nit: you could remove the `else` clause and move the `args.Cancel` assignment outside the if to `!_cts.IsCancellationRequested`.

File: src/Microsoft.DotNet.Watcher.Tools/CommandLineOptions.cs (hunk 1 above)
Comment (natemcmaster) [inline src/Microsoft.DotNet.Watcher.Tools/CommandLineOptions.cs]: thanks. Updated.

**Pull request merged**